pkg/eestream/rs_test.go

```diff
@@ -99,6 +101,188 @@ func TestRSErrors(t *testing.T) {
+		{6 * 1024, 1024, 3, 7, 0, false},
+		{6 * 1024, 1024, 3, 7, 1, false},
+		{6 * 1024, 1024, 3, 7, 2, false},
+		{6 * 1024, 1024, 3, 7, 3, false},
+		{6 * 1024, 1024, 3, 7, 4, false},
+		{6 * 1024, 1024, 3, 7, 5, true},
+		{6 * 1024, 1024, 3, 7, 6, true},
+		{6 * 1024, 1024, 3, 7, 7, true},
+	} {
+		errTag := fmt.Sprintf("Test case #%d", i)
+		data := randData(tt.dataSize)
+		fc, err := infectious.NewFEC(tt.required, tt.total)
+		if !assert.NoError(t, err, errTag) {
+			continue
+		}
+		rs := NewRSScheme(fc, tt.blockSize)
+		readers := EncodeReader(bytes.NewReader(data), rs)
+		// read all readers in []byte buffers to avoid deadlock if later
+		// we don't read in parallel from all of them
+		pieces, err := readAll(readers)
+		if !assert.NoError(t, err, errTag) {
+			continue
+		}
+		readerMap := make(map[int]io.ReadCloser, len(readers))
+		// some readers will return EOF earlier
+		for i := 0; i < tt.problematic; i++ {
+			readerMap[i] = readcloser.LimitReadCloser(
+				ioutil.NopCloser(bytes.NewReader(pieces[i])),
+				int64(rand.Intn(tt.dataSize)))
+		}
+		// the rest will operate normally
+		for i := tt.problematic; i < tt.total; i++ {
+			readerMap[i] = ioutil.NopCloser(bytes.NewReader(pieces[i]))
+		}
+		data2, err := ioutil.ReadAll(DecodeReaders(readerMap, rs))
+		if !tt.fail && assert.NoError(t, err, errTag) {
+			assert.Equal(t, data, data2, errTag)
+		}
+	}
+}
+
+// Some pieces will read EOF later than expected
+func TestRSLateEOF(t *testing.T) {
+	for i, tt := range []struct {
+		dataSize    int
+		blockSize   int
+		required    int
+		total       int
+		problematic int
+		fail        bool
+	}{
+		{4 * 1024, 1024, 1, 1, 0, false},
+		{4 * 1024, 1024, 1, 1, 1, true},
+		{4 * 1024, 1024, 1, 2, 0, false},
+		{4 * 1024, 1024, 1, 2, 1, false},
+		{4 * 1024, 1024, 1, 2, 2, true},
+		{4 * 1024, 1024, 2, 4, 0, false},
+		{4 * 1024, 1024, 2, 4, 1, false},
+		{4 * 1024, 1024, 2, 4, 2, false},
+		{4 * 1024, 1024, 2, 4, 3, true},
+		{4 * 1024, 1024, 2, 4, 4, true},
+		{6 * 1024, 1024, 3, 7, 0, false},
+		{6 * 1024, 1024, 3, 7, 1, false},
+		{6 * 1024, 1024, 3, 7, 2, false},
+		{6 * 1024, 1024, 3, 7, 3, false},
+		{6 * 1024, 1024, 3, 7, 4, false},
+		{6 * 1024, 1024, 3, 7, 5, true},
+		{6 * 1024, 1024, 3, 7, 6, true},
+		{6 * 1024, 1024, 3, 7, 7, true},
+	} {
+		errTag := fmt.Sprintf("Test case #%d", i)
+		data := randData(tt.dataSize)
+		fc, err := infectious.NewFEC(tt.required, tt.total)
+		if !assert.NoError(t, err, errTag) {
+			continue
+		}
+		rs := NewRSScheme(fc, tt.blockSize)
+		readers := EncodeReader(bytes.NewReader(data), rs)
+		// read all readers in []byte buffers to avoid deadlock if later
+		// we don't read in parallel from all of them
+		pieces, err := readAll(readers)
+		if !assert.NoError(t, err, errTag) {
+			continue
+		}
+		readerMap := make(map[int]io.ReadCloser, len(readers))
+		// some readers will return EOF later
+		for i := 0; i < tt.problematic; i++ {
+			readerMap[i] = readcloser.LimitReadCloser(
```

Comment:
This is wrong. I cannot use LimitReadCloser to return more data than available in the provided reader.